# Data Cleaning

In [ ]:
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np
import seaborn as sns
from sklearn.model_selection import train_test_split

# Scalling untuk MinMax Scaller
from sklearn.preprocessing import MinMaxScaler
# Set the maximum number of columns and rows to display to a large number
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

# B. Outlier Handling

### Outlier Handling : InterQuartile Handling

In [ ]:
# untuk plot Q-Q
import scipy.stats as stats

### Assignment 1.a :
1. lakukan looping untuk melihat plot distribusi data dan outlier pada kolom-kolom : ['MedInc', 'HouseAge', 'AveRooms','AveBedrms', 'AveOccup'] di california_dataset.csv
2. Daftar nama kolom yang Anda cek, kolom_kolom = ['MedInc', 'HouseAge', 'AveRooms', 'AveBedrms', 'AveOccup']
3. Lakukan untuk tujuan machine learning dan juga exploratory data analysis.

### untuk tujuan machine learning

In [ ]:
df_california = pd.read_csv('california_dataset.csv')
df_california.head()

FileNotFoundError: [Errno 2] No such file or directory: 'california_dataset.csv'

In [ ]:
from sklearn.model_selection import train_test_split
train_clfr, test_clfr = train_test_split(df_california, test_size=0.2, random_state=42)

In [ ]:
train_clfr.describe()

In [ ]:
df_california.columns

In [ ]:
kolom_kolom = ['MedInc', 'HouseAge', 'AveRooms', 'AveBedrms', 'AveOccup']

In [ ]:
def check_plot(df, variable):

  plt.figure(figsize=(16, 4))

  #plot histogram
  plt.subplot(1,3,1)
  sns.histplot(df[variable], bins=30)
  plt.title(f'Histogram of {variable}')
  plt.xlabel(variable)
  plt.ylabel('Frequency')

  #plot Q-Q
  plt.subplot(1,3,2)
  stats.probplot(df[variable], dist="norm", plot=plt)
  plt.ylabel('Variable quantiles')

  #plot boxplot
  plt.subplot(1,3,3)
  sns.boxplot(y=df[variable])
  plt.title(f'Box Plot of {variable}')
  plt.show()


for kolom in kolom_kolom:
  print(f"Visualisasi Kolom: {kolom}")
  check_plot(train_clfr, kolom)

Insight:
1. Rata-rata menunjukkan pola distribusi yang skewed right terlihat pada kolom Averooms, MedInc, AveBedrms, AveOccup.
2. Terdapat banyak outlier di kolom-kolom tersebut, yang terlihat jelas di Box Plot.
3. Distribusi data secara umum tidak mengikuti distribusi normal yang terlihat pada Probability Plot.

### untuk exploratory data analysis

In [ ]:
data_clfr = df_california.copy()

In [ ]:
for kolom in kolom_kolom:
  print(f"Visualisasi Kolom: {kolom} (Exploratory Data Analysis)")
  check_plot(data_clfr, kolom)

Insight:
1. Rata-rata menunjukkan pola distribusi yang skewed right terlihat pada kolom Averooms, MedInc, AveBedrms, AveOccup.
2. Terdapat banyak outlier di kolom-kolom tersebut, yang terlihat jelas di Box Plot.
3. Distribusi data secara umum tidak mengikuti distribusi normal terlihat pada Probability Plot.


1. Lakukan handling outlier pilih 1 kolom saja yang di handling dengan IQR setelah mengecek distribusinya
2. Dan cek distribusinya kembali
3. untuk tujuan machine learning dan juga exploratory data analysis.

In [ ]:
### untuk tujuan machine learning### Assignment 1.b Lakukan handling outlier pilih 1 kolom saja yang di handling dengan IQR setelah mengecek distribusinya
### Dan cek distribusinya kembali

### untuk tujuan machine learning

In [ ]:
def find_outlier_boundary(df, variable):

    # Menghitung batas luar yang merupakan outlier

    IQR = df[variable].quantile(0.75) - df[variable].quantile(0.25)

    lower_boundary = df[variable].quantile(0.25) - (IQR * 1.5)
    upper_boundary = df[variable].quantile(0.75) + (IQR * 1.5)

    return upper_boundary, lower_boundary

In [ ]:
# batas untuk AveBedrms
occup_upper_limit, occup_lower_limit = find_outlier_boundary(train_clfr, 'AveBedrms')
occup_upper_limit, occup_lower_limit

In [ ]:
# Trimming atau drop value outlier -> pengurangan baris
# Trimming outliers from the training data
train_clfr = train_clfr[(train_clfr['AveBedrms'] <= occup_upper_limit) & (train_clfr['AveBedrms'] >= occup_lower_limit)]

In [ ]:
check_plot(train_clfr, 'AveBedrms')

Insight:
Hasil yang didapat setelah dilakukan handling outlier pada kolom AveBedrms pada data train yaitu:
1. Histogram menunjukan distribusi data hampir simetris dengan sedikit skewness ke kanan (skewed right).
2. Probability Plot menunjukan garis data (titik-titik biru) mengikuti garis lurus merah dengan baik di bagian tengah, meskipun ada sedikit penyimpangan di bagian ujung kanan atas hal ini menunjukkan bahwa distribusi hampir normal dengan sedikit skewness positif (skewed right).
3. Box plot menunjukkan distribusi data cukup seimbang, hampir tidak ada outlier ekstrem di luar whiske.

### untuk exploratory data analysis

In [ ]:
full_occup_upper_limit, full_occup_lower_limit = find_outlier_boundary(data_clfr,'AveBedrms')
full_occup_upper_limit, full_occup_lower_limit

In [ ]:
data_clfr = data_clfr[(data_clfr['AveBedrms'] <= full_occup_upper_limit) & (data_clfr['AveBedrms'] >= full_occup_lower_limit)]

In [ ]:
check_plot(data_clfr, 'AveBedrms')

Insight:
Hasil yang didapat setelah dilakukan handling outlier kolom AveBedrms pada data_clfr untuk Exploratory data analysis yaitu:
1. Histogram menunjukan distribusi data hampir simetris dengan sedikit skewness ke kanan (skewed right).
2. Probability Plot menunjukan garis data (titik-titik biru) mengikuti garis lurus merah dengan baik di bagian tengah, meskipun ada sedikit penyimpangan di bagian ujung kanan atas hal ini menunjukkan bahwa distribusi hampir normal dengan sedikit skewness positif (skewed right).
3. Box plot menunjukkan distribusi data cukup seimbang, hampir tidak ada outlier ekstrem di luar whiske.

# C. Missing Value Handling
Missing Value Merupakan nilai yang hilang di suatu kolom, namun tidak hanya sebatas itu. Untuk nilai yang tidak sesuai tapi muncul didalam suatu kolom tertentu juga bisa dianggap sebagai missing value. Misal tipenya kategorikal tapi isinya ada yang berupa angka dan juga sebaliknya.

### Aturan missing value handling :
1. jika tipe datanya itu **numerik** -> handling menggunakan **median** (robust / tahan terhadap outlier)
2. jika tipe datanya itu **kategorik / object / string** -> handling menggunakan **mode / modus**

## Missing Value Handling :
1. Cek persentase missing value hanya pada kolom Headquarters di company.csv, Cek jenis missing valuenya apakah ada atau tidak ?
2. apakah di drop atau tidak (gunakan syarat drop kolom jika > 20%)
3. jika tidak maka handling, dan dengan cara apa handling nilainya, median atau modus ?
4. Lakukan untuk tujuan machine learning dan juga exploratory data analysis.

In [ ]:
df_company = pd.read_csv('company.csv')

In [ ]:
kolom = 'Headquarters'
print(f"============= {kolom} =================")
display(df_company[kolom].value_counts())
print()

============= Headquarters =================


Headquarters
New York, NY                 33
-1                           31
San Francisco, CA            31
Chicago, IL                  23
Boston, MA                   19
Reston, VA                   14
Mc Lean, VA                  13
Westminster, CO              12
Cambridge, United Kingdom    10
Cambridge, MA                10
Vienna, VA                    8
Denver, CO                    7
Herndon, VA                   7
Washington, DC                7
Woodbine, MD                  7
Saint Louis, MO               7
Bedford, MA                   7
Arlington, VA                 7
San Jose, CA                  7
San Antonio, TX               6
Mountain View, CA             6
Basel, Switzerland            6
San Rafael, CA                6
Chantilly, VA                 6
Fairfax, VA                   5
Houston, TX                   5
Pittsburgh, PA                5
Bethesda, MD                  5
Rockville, MD                 5
Los Angeles, CA               5
Springfield, MA               5
Bellevue, WA                  5
Princeton, NJ                 5
Annapolis Junction, MD        4
Cedar Rapids, IA              4
South San Francisco, CA       4
McLean, VA                    4
Falls Church, VA              4
Orlando, FL                   4
Pleasanton, CA                4
Raleigh, NC                   4
Aliso Viejo, CA               4
Santa Clara, CA               4
Santa Cruz, CA                4
San Diego, CA                 4
Austin, TX                    4
Palo Alto, CA                 4
Memphis, TN                   4
Indianapolis, IN              3
Paris, France                 3
Naperville, IL                3
Atlanta, GA                   3
Hillsborough, NJ              3
Sunnyvale, CA                 3
Minneapolis, MN               3
Laurel, MD                    3
Columbus, OH                  3
Winter Park, FL               3
Alexandria, VA                3
Lincoln, NE                   3
Jersey City, NJ               3
Woburn, MA                    3
Carson, CA                    3
Folsom, CA                    3
Westlake Village, CA          3
Richland, WA                  3
Seattle, WA                   3
Maple Plain, MN               3
Hartford, CT                  3
Fort Worth, TX                2
Lorton, VA                    2
Itasca, IL                    2
Carle Place, NY               2
Fairmont, WV                  2
Kent, WA                      2
Coconut Creek, FL             2
Cupertino, CA                 2
Waltham, MA                   2
Baltimore, MD                 2
Nashville, TN                 2
Plainsboro, NJ                2
Toronto, Canada               2
Germantown, MD                2
Kansas City, MO               2
Winston-Salem, NC             2
Huntsville, AL                2
New York, 061                 2
Wilmington, MA                2
Danville, CA                  2
Columbia, MD                  2
San Carlos, CA                2
North Brunswick, NJ           2
Edison, NJ                    2
Kent, OH                      2
Leesburg, VA                  2
Saint Paul, MN                2
London, United Kingdom        2
Cincinnati, OH                2
Brentford, United Kingdom     2
Fremont, CA                   2
Irvine, CA                    2
Somerset, NJ                  2
Scotts Valley, CA             2
Lynchburg, VA                 2
Stamford, CT                  2
San Mateo, CA                 2
Holmdel, NJ                   2
South Jordan, UT              2
Emeryville, CA                2
Hamilton, Bermuda             2
San Ramon, CA                 2
OSAKA, Japan                  2
Singapore, Singapore          2
Santa Monica, CA              2
Sheboygan, WI                 2
Lexington, MA                 2
Menlo Park, CA                2
Mays Landing, NJ              2
Northbrook, IL                2
Santa Barbara, CA             2
Lehi, UT                      2
Chevy Chase, MD               2
Greenville, SC                2
Saxonburg, PA                 1
Livonia, MI        

insight:

Terdapat Missing value pada kolom headquarters dengan nilai **-1**

IDENTIFYING PERCENTAGE (%) MISSING DATA

In [ ]:
#Tulis jawaban missing value handling kolom Headquarters disini
100 * len(df_company[(df_company['Headquarters'] == '-1') | (df_company['Headquarters'].isna())]) / len(df_company)

4.613095238095238

Insight:

Menunjukkan bahwa 4.61% dari total baris dalam DataFrame df_company memiliki nilai tidak valid di kolom **'Headquarters'** berupa -1 atau NaN, maka dilakukan handling dengan menggunakan modus

### untuk tujuan machine learning

In [ ]:
from sklearn.model_selection import train_test_split
train_company, test_company = train_test_split(df_company, test_size = 0.2, random_state=42)
#train : test = 80:20

In [ ]:
#Mode imputation Untuk Headquarters
train_mode_headquarters = train_company[~train_company['Headquarters'].isin(['-1'])]['Headquarters'].mode()[0]
train_mode_headquarters

insight:

Menunjukkan bahwa 'New York, NY' adalah modus (nilai yang paling sering muncul) pada kolom 'Headquarters' dalam data train_company

In [ ]:
#implement missing value imputation to train and test, from Headquarters mode value in train data
train_company['Headquarters'] = train_company['Headquarters'].apply(lambda x: train_mode_headquarters if x in ['-1'] else x)
test_company['Headquarters'] = test_company['Headquarters'].apply(lambda x: train_mode_headquarters if x in ['-1'] else x)

In [ ]:
train_company['Headquarters'].isna().sum()

In [ ]:
test_company['Headquarters'].isna().sum()

Insight:

Tidak terdapat Missing Value pada data train dan test

### untuk exploratory data analysis

In [ ]:
#Tulis jawaban missing value handling kolom Headquarters disini
data_company = df_company.copy()

In [ ]:
#Mode imputation Untuk Headquarters
mode_headquarters = data_company[~data_company['Headquarters'].isin(['-1'])]['Headquarters'].mode()[0]
mode_headquarters

'New York, NY'

insight:

Menunjukkan bahwa 'New York, NY' adalah modus (nilai yang paling sering muncul) pada kolom 'Headquarters' dalam data_company

In [ ]:
#implement missing value imputation to data_company from headquarters
data_company['Headquarters'] = data_company['Headquarters'].apply(lambda x: mode_headquarters if x in ['-1'] else x)

In [ ]:
data_company['Headquarters'].isna().sum()

0

Insight:

Tidak Terdapat Missing Value pada data_company untuk exploratory data analysis

# D. Encoding
Definisi : Mengubah tipe kategorikal menjadi numerikal dengan pendekatan aturan encoding untuk persiapan sebelum dilakukannya machine learning modelling. Teknik Encoding yang umum dipakai :
1. One Hot Encoding (OHE) : Lakukan pada kolom gender
2. Label Encoding / Ordinal Encoding : Lakukan pada kolom Yes dan No, Jika Yes menjadi 1 , Jika No menjadi 0. Sebagai Contohnya di kolom  Partner, Dependents.
3. Mean Encoding : Contract

In [ ]:
df_churn = pd.read_csv('Telco-Customer-Churn.csv')
# Mengganti "Yes" menjadi 1 dan "No" menjadi 0 dalam kolom "Churn"
df_churn['Churn'] = df_churn['Churn'].replace({'Yes': 1, 'No': 0})
df_churn.head()

## Assignment 3.a Label Encoder :

0. Lakukan label encoder pada 'Partner', 'Dependents' juga sama seperti yang telah dibahas dikelas, **lalu anda wajib melakukan** :
1. Pilih kolom-kolom yang terindikasi bersifat Data Ordinal.
2. Ganti nilai yang tidak sesuai : No internet service menjadi No lalu validasi dengan value_counts
3. Kemudian lakukan label Encoder
4. Lakukan untuk tujuan machine learning dan juga exploratory data analysis.
5. Notes : gunakan Telco-Customer-Churn.csv

In [ ]:
#Tulis jawaban
from sklearn.preprocessing import LabelEncoder

In [ ]:
df_churn.info()

In [ ]:
df_churn.columns

In [ ]:
#mengecek nilai-nilai di kolom kategorikal
custom_columns = ['Partner','Dependents','PhoneService','MultipleLines',
                  'InternetService','OnlineSecurity','OnlineBackup','DeviceProtection'
                  ,'TechSupport','StreamingTV','StreamingMovies','Contract','PaperlessBilling','PaymentMethod','Churn']
for column in custom_columns:
    print(f"============= {column} =================")
    display(df_churn[column].value_counts())
    print()

In [ ]:
#Lakukan normalisasi nilai terlebih dahulu
df_churn['StreamingMovies'] = df_churn['StreamingMovies'].replace({'No internet service':'No'})
df_churn['StreamingTV'] = df_churn['StreamingTV'].replace({'No internet service':'No'})
df_churn['TechSupport'] = df_churn['TechSupport'].replace({'No internet service':'No'})
df_churn['DeviceProtection'] = df_churn['DeviceProtection'].replace({'No internet service':'No'})
df_churn['OnlineBackup'] = df_churn['OnlineBackup'].replace({'No internet service':'No'})
df_churn['OnlineSecurity'] = df_churn['OnlineSecurity'].replace({'No internet service':'No'})
df_churn['MultipleLines'] = df_churn['MultipleLines'].replace({'No phone service':'No'})

In [ ]:
#mengecek kembali nilai-nilai di kolom kategorikal
for column in custom_columns:
    print(f"============= {column} =================")
    display(df_churn[column].value_counts())
    print()

### untuk tujuan machine learning

In [ ]:
from sklearn.model_selection import train_test_split
train_churn, test_churn = train_test_split(df_churn, test_size = 0.2, random_state=42)

In [ ]:
from sklearn.preprocessing import LabelEncoder

# Kolom-kolom yang ingin di-label encode
columns_to_encode = ['Partner', 'Dependents', 'PhoneService', 'MultipleLines' , 'OnlineSecurity', 'OnlineBackup', 'DeviceProtection', 'TechSupport', 'StreamingTV', 'StreamingMovies', 'PaperlessBilling']

# Inisialisasi LabelEncoder
label_encoder = LabelEncoder()

In [ ]:
# Menerapkan label encoding ke data pelatihan (train)
for col in columns_to_encode:
    train_churn[col] = label_encoder.fit_transform(train_churn[col])

In [ ]:
# Menerapkan label encoding yang sama ke data uji (test)
for col in columns_to_encode:
    test_churn[col] = label_encoder.transform(test_churn[col])

In [ ]:
train_churn.head()

In [ ]:
test_churn.head()

### untuk exploratory data analysis

In [ ]:
data_churn = df_churn.copy()

In [ ]:
# Kolom-kolom yang ingin di-label encode
columns_to_encode = ['Partner', 'Dependents','PhoneService', 'MultipleLines' , 'OnlineSecurity', 'OnlineBackup', 'DeviceProtection', 'TechSupport', 'StreamingTV', 'StreamingMovies', 'PaperlessBilling']

# Inisialisasi LabelEncoder
labelEncoder = LabelEncoder()

In [ ]:
# Menerapkan label encoding ke data pelatihan (train)
for col in columns_to_encode:
    data_churn[col] = labelEncoder.fit_transform(data_churn[col])
#fit dilakukan di data train dan implementasikan ke train, kemudian lakukan transform ke data test. Untuk mencegah data leakage

In [ ]:
data_churn.head()

##  Mean Encoding
1. Lakukan kolom Mean Encoding pada InternetService

### untuk tujuan machine learning

In [ ]:
# Tulis jawaban
train_churn.head()

In [ ]:
dict_internetservice = train_churn.groupby(['InternetService'])['Churn'].mean().to_dict()
dict_internetservice

Insight:
1. DSL memiliki rata-rata tingkat Churn sekitar 19.18%, yang berarti sekitar 19.18% dari Customer dengan kategori Churn.
2. Fiber optic memiliki rata-rata tingkat Churn sekitar 41.55%, yang berarti sekitar 41.55% dari Customer dengan kategori Churn.
3. No memiliki rata-rata tingkat Churn sekitar 7.66%, yang berarti sekitar 7.66% dari Customer dengan kategori Churn.

In [ ]:
# Mengubah nilai kategorikal di data train dan test pada kolom InternetService menggunakan nilai dari dict_internetservice
# untuk mencegah data leakage
train_churn['InternetService'] = train_churn['InternetService'].map(dict_internetservice)
test_churn['InternetService'] = test_churn['InternetService'].map(dict_internetservice)

In [ ]:
train_churn.head()

In [ ]:
test_churn.head()

### untuk exploratory data analysis

In [ ]:
dict_encoding_eda_internetservice = data_churn.groupby(['InternetService'])['Churn'].mean().to_dict()
dict_encoding_eda_internetservice

Insight:
1. DSL memiliki rata-rata tingkat Churn sekitar 18.95%, yang berarti sekitar 18.95% dari Customer dengan kategori Churn.
2. Fiber optic memiliki rata-rata tingkat Churn sekitar 41.89%, yang berarti sekitar 41.89% dari Customer dengan kategori Churn.
3. No memiliki rata-rata tingkat Churn sekitar 7.40%, yang berarti sekitar 7.40% dari Customer dengan kategori Churn.

In [ ]:
# Mengubah nilai kategorikal di data_churn pada kolom InternetService menggunakan nilai dari dict_encoding_eda_internetservice
# untuk mencegah data leakage
data_churn['InternetService'] = data_churn['InternetService'].map(dict_encoding_eda_internetservice)

In [ ]:
data_churn.head()